### PR TITLE
feat: Support script for sharing all cases from single LA

### PR DESCRIPTION
**Pull request by tomaszpowroznik**

### CODE REVIEW

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/support/ShareCaseController.java [hunk 1]
```diff
@@ -45,9 +56,52 @@ public void shareCase(@PathVariable("caseId") String caseId, @RequestBody Users
...
     }
 
+    @PostMapping(value = "/support/cases/share", consumes = APPLICATION_JSON_VALUE)
+    @Secured("caseworker-publiclaw-systemupdate")
+    public void shareCases(@Valid @RequestBody ShareLocalAuthorityCases shareCases) {
+
+        final String query = new JSONObject()
+            .put("from", 0)
+            .put("size", 1000)
+            .put("query", of("match", of("data.caseLocalAuthority", shareCases.localAuthority)))
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/support/ShareCaseController.java:66]: Would it be better to use a bool query here? https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-bool-query.html

Comment (lewisbirks) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/support/ShareCaseController.java:66]: something like
```json
{
  "query": {
    "bool" : {
      "must" : {
        "term" : { "data.caseLocalAuthority" : "LA" }
      }
    }
  }
}
```

File: scripts/shareLocalAuthorityCases.sh [hunk 2]
```diff
@@ -0,0 +1,50 @@
+#!/usr/bin/env bash
+
+set -eu
+
+main() {
+    checkArgs "$@"
+
+    user=${1}
+    password=${2}
+    localAuthority=${3}
```
Comment (lewisbirks) [inline scripts/shareLocalAuthorityCases.sh:10]: Don't see this being used in this method

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/support/ShareCaseController.java [hunk 3]
```diff
@@ -45,9 +56,52 @@ public void shareCase(@PathVariable("caseId") String caseId, @RequestBody Users
...
+        final List<CaseDetails> casesDetails = coreCaseDataService.searchCases(CASE_TYPE, query).stream()
+            //`Extra precaution as ES match does not guarantee strict equality
+            .filter(caseDetails -> shareCases.localAuthority.equals(caseDetails.getData().get("caseLocalAuthority")))
+            .collect(Collectors.toList());
+
+        log.info("Share case. Found {} case(s) for '{}' LA", casesDetails.size(), shareCases.localAuthority);
+
+        final String serviceToken = authTokenGenerator.generate();
+
+        casesDetails.stream()
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/support/ShareCaseController.java:78]: This first stream isn't really required

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/support/ShareCaseController.java [hunk 4]
```diff
@@ -45,9 +56,52 @@ public void shareCase(@PathVariable("caseId") String caseId, @RequestBody Users
...
+            .forEach(caseDetails ->
+                shareCases.usersIds.forEach(userId -> {
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/support/ShareCaseController.java:80]: This part is very similar to the above method. Maybe there could be a private method 
```java
    private void shareCase(List<String> ids, String caseId, String serviceToken) {
        ids.forEach(id -> {
            caseUser.updateCaseRolesForUser(
                requestData.authorisation(),
                serviceToken,
                caseId,
                id,
                new CaseUser(id, CASE_ROLES)
            );
            log.info("Share case. User {} granted access to {}", id, caseId);
        });
    } 
```

Comment (lewisbirks) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/support/ShareCaseController.java:80]: Then the two controller methods can use that

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/support/ShareCaseController.java (hunk 3 above)
Comment (tomaszpowroznik) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/support/ShareCaseController.java]: done

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/support/ShareCaseController.java (hunk 4 above)
Comment (tomaszpowroznik) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/support/ShareCaseController.java]: done

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/support/ShareCaseController.java [hunk 5]
```diff
@@ -11,43 +13,92 @@
...
+        final String query = new JSONObject()
+            .put("from", 0)
+            .put("size", 1000)
+            .put("query", of("match", of("data.caseLocalAuthority", shareCases.localAuthority)))
+            .toString();
 
+        final List<String> caseIds = coreCaseDataService.searchCases(CASE_TYPE, query).stream()
+            //`Extra precaution as ES match does not guarantee strict equality
+            .filter(caseDetails -> shareCases.localAuthority.equals(caseDetails.getData()
+                .getOrDefault("caseLocalAuthority", null)))
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/support/ShareCaseController.java:63]: won't get return null by default anyway?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/support/ShareCaseController.java (hunk 1 above)
Comment (lewisbirks) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/support/ShareCaseController.java:57]: fair enough, was just copying their example to our fields

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/support/ShareCaseControllerTest.java [hunk 6]
```diff
@@ -115,4 +214,31 @@ private MvcResult shareCase(final String caseId, String payload) throws Exceptio
...
+                .content(payload)
+                .header("authorization", USER_AUTH_TOKEN)
+                .contentType(APPLICATION_JSON))
+            .andReturn();
+    }
+
+    private static CaseDetails caseDetails(String localAuthority) {
+        Map<String, Object> caseData = new HashMap<>();
+
+        Optional.ofNullable(localAuthority).ifPresent(la -> caseData.put("caseLocalAuthority", la));
```
Comment (lewisbirks) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/support/ShareCaseControllerTest.java:237]: Would a simple if not be more readable here?

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/support/ShareCaseControllerTest.java [hunk 7]
```diff
@@ -50,51 +64,136 @@
...
+
+            CaseDetails case1 = caseDetails(localAuthority);
+            CaseDetails case2 = caseDetails(localAuthority2);
+            CaseDetails case3 = caseDetails(null);
+            CaseDetails case4 = caseDetails("");
+            CaseDetails case5 = caseDetails(localAuthority);
+
+            given(authTokenGenerator.generate()).willReturn(SERVICE_AUTH_TOKEN);
+            given(coreCaseDataService.searchCases(any(), queryCaptor.capture()))
+                .willReturn(List.of(case1, case2, case3, case4, case5));
```
Comment (lewisbirks) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/support/ShareCaseControllerTest.java:94]: should case 2 be returned here?

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/support/ShareCaseControllerTest.java (hunk 6 above)
Comment (tomaszpowroznik) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/support/ShareCaseControllerTest.java]: yeah, done

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/support/ShareCaseControllerTest.java (hunk 7 above)
Comment (tomaszpowroznik) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/support/ShareCaseControllerTest.java:94]: it is to test the additional check, changed a bit test data to make it more realistic

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/support/ShareCaseController.java (hunk 5 above)
Comment (tomaszpowroznik) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/support/ShareCaseController.java]: of course,  fixed

**Review by lewisbirks — APPROVED** [review]

**Pull request merged**